client/components/tinymce/plugins/wpcom-view/wpvideo-view.jsx

```diff
@@ -0,0 +1,146 @@
+/**
+ * External dependencies
+ */
+import React, { Component, PropTypes } from 'react';
+import { connect } from 'react-redux';
+import omitBy from 'lodash/omitBy';
+import pick from 'lodash/pick';
+import QueryString from 'querystring';
+import tinymce from 'tinymce/tinymce';
+
+/**
+ * Internal dependencies
+ */
+import shortcodeUtils from 'lib/shortcode';
+import { getVideo } from 'state/videos/selectors';
+import QueryVideo from 'components/data/query-video';
+
+class WpVideoView extends Component {
+
+	static match( content ) {
+		const match = shortcodeUtils.next( 'wpvideo', content );
+
+		if ( match ) {
+			return {
+				index: match.index,
+				content: match.content,
+				options: {
+					shortcode: match.shortcode
+				}
+			};
+		}
+	}
+
+	static serialize( content ) {
+		return encodeURIComponent( content );
+	}
+
+	constrainVideoDimensions( shortcodeWidthAttribute, shortcodeHeightAttribute ) {
+		const defaultWidth = 640;
+		const defaultAspectRatio = 16 / 9;
+		let { width, height } = this.props.video
+			? this.props.video
+			: {
+				width: defaultWidth,
+				height: defaultWidth / defaultAspectRatio
+			};
+		const aspectRatio = width / height;
+
+		if ( shortcodeWidthAttribute && ! shortcodeHeightAttribute ) {
+			width = shortcodeWidthAttribute;
+			height = shortcodeWidthAttribute / aspectRatio;
+		} else if ( ! shortcodeWidthAttribute && shortcodeHeightAttribute ) {
+			width = shortcodeHeightAttribute * aspectRatio;
+			height = shortcodeHeightAttribute;
+		} else if ( shortcodeWidthAttribute && shortcodeHeightAttribute ) {
+			const definedAspectRatio = shortcodeWidthAttribute / shortcodeHeightAttribute;
```

Comment:
Good thinking. We should keep the user-defined width and height.

Even though it may result in an aspect ratio that might not be the same as the Storage Aspect Ration (SAR) the user-defined dimensions should take precedence (the user knows what they are doing, we should not be too smart here).

In some cases the video might have a different Display Aspect Ratio (DAR) than SAR, so this gives the user a chance to get the ratio right.